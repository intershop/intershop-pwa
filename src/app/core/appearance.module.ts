import { NgModule } from '@angular/core';
import { NgbPopoverConfig } from '@ng-bootstrap/ng-bootstrap';
import { ToastrModule } from 'ngx-toastr';

import { IconModule } from './icon.module';
import { ThemeService } from './utils/theme/theme.service';

@NgModule({
  imports: [
    IconModule,
    ToastrModule.forRoot({
      closeButton: true,
      timeOut: 3000,
      positionClass: 'toast-top-full-width', // toast-top-center
      preventDuplicates: true,
    }),
  ],
})
export class AppearanceModule {
  constructor(popoverConfig: NgbPopoverConfig, themeService: ThemeService) {
    popoverConfig.placement = 'top';
    popoverConfig.triggers = 'hover';
    popoverConfig.container = 'body';

    themeService.init();
  }
}
