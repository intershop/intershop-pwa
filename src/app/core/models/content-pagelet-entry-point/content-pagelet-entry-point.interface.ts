import { ContentConfigurationParameterData } from 'ish-core/models/content-configuration-parameter/content-configuration-parameter.interface';
import { ContentPageletData } from 'ish-core/models/content-pagelet/content-pagelet.interface';
import { Link } from 'ish-core/models/link/link.model';

export interface ContentPageletEntryPointData {
  id: string;
  definitionQualifiedName: string;
  displayName: string;
  domain: string;
  resourceSetId: string;
  pagelets: ContentPageletData[];
  configurationParameters?: { [name: string]: ContentConfigurationParameterData };
  link: Link;
}
