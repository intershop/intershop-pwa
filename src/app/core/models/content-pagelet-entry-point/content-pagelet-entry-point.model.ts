import { ContentConfigurationParameters } from 'ish-core/models/content-configuration-parameter/content-configuration-parameter.mapper';

export interface ContentPageletEntryPoint {
  id: string;
  definitionQualifiedName: string;
  displayName: string;
  domain: string;
  resourceSetId: string;
  pageletIDs?: string[];
  configurationParameters?: ContentConfigurationParameters;
  clientId?: string;
}
