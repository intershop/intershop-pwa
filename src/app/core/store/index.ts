export * from './actions';
export * from './effects';
export * from './reducers';
export * from './selectors';
export * from './serializers/router.serializers';
