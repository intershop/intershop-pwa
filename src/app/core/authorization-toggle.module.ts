import { ModuleWithProviders, NgModule } from '@angular/core';
import { ReplaySubject } from 'rxjs';
import { map } from 'rxjs/operators';

import { AuthorizationToggleDirective } from './directives/authorization-toggle.directive';
import { AuthorizationToggleService, checkPermission } from './utils/authorization-toggle/authorization-toggle.service';
import { whenTruthy } from './utils/operators';

@NgModule({
  declarations: [AuthorizationToggleDirective],
  exports: [AuthorizationToggleDirective],
})
export class AuthorizationToggleModule {
  private static permissions = new ReplaySubject<string[]>(1);

  static forTesting(...permissions: string[]): ModuleWithProviders<AuthorizationToggleModule> {
    AuthorizationToggleModule.switchTestingPermissions(...permissions);
    return {
      ngModule: AuthorizationToggleModule,
      providers: [
        {
          provide: AuthorizationToggleService,
          useValue: {
            isAuthorizedTo: (permission: string) =>
              AuthorizationToggleModule.permissions.pipe(
                whenTruthy(),
                map(perms => checkPermission(perms, permission))
              ),
          },
        },
      ],
    };
  }

  static switchTestingPermissions(...permissions: string[]) {
    AuthorizationToggleModule.permissions.next(permissions);
  }
}

export { AuthorizationToggleService } from './utils/authorization-toggle/authorization-toggle.service';
export { AuthorizationToggleGuard } from './guards/authorization-toggle.guard';
