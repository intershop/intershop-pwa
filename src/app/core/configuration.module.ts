import { NgModule } from '@angular/core';

import { SPECIAL_HTTP_ERROR_HANDLER } from './interceptors/icm-error-mapper.interceptor';
import { createPaymentErrorHandler } from './utils/http-error/create-payment.error-handler';
import { editPasswordErrorHandler } from './utils/http-error/edit-password.error-handler';
import { LoginUserErrorHandler } from './utils/http-error/login-user.error-handler';
import { requestReminderErrorHandler } from './utils/http-error/request-reminder.error-handler';
import { updatePasswordErrorHandler } from './utils/http-error/update-password.error-handler';

@NgModule({
  providers: [
    { provide: SPECIAL_HTTP_ERROR_HANDLER, useValue: updatePasswordErrorHandler, multi: true },
    { provide: SPECIAL_HTTP_ERROR_HANDLER, useClass: LoginUserErrorHandler, multi: true },
    { provide: SPECIAL_HTTP_ERROR_HANDLER, useValue: requestReminderErrorHandler, multi: true },
    { provide: SPECIAL_HTTP_ERROR_HANDLER, useValue: editPasswordErrorHandler, multi: true },
    { provide: SPECIAL_HTTP_ERROR_HANDLER, useValue: createPaymentErrorHandler, multi: true },
  ],
})
export class ConfigurationModule {}
