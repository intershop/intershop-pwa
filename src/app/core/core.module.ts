import { APP_BASE_HREF, PlatformLocation } from '@angular/common';
import { HTTP_INTERCEPTORS, HttpClientModule } from '@angular/common/http';
import { ErrorHandler, NgModule, Optional, SkipSelf } from '@angular/core';
import { ServiceWorkerModule } from '@angular/service-worker';
import { TranslateModule } from '@ngx-translate/core';

import { AppearanceModule } from './appearance.module';
import { ConfigurationModule } from './configuration.module';
import { FeatureToggleModule } from './feature-toggle.module';
import { IdentityProviderModule } from './identity-provider.module';
import { ICMErrorMapperInterceptor } from './interceptors/icm-error-mapper.interceptor';
import { IdentityProviderInterceptor } from './interceptors/identity-provider.interceptor';
import { MockInterceptor } from './interceptors/mock.interceptor';
import { InternationalizationModule } from './internationalization.module';
import { StateManagementModule } from './state-management.module';
import { DefaultErrorHandler } from './utils/default-error-handler';

@NgModule({
  imports: [
    AppearanceModule,
    ConfigurationModule,
    FeatureToggleModule,
    HttpClientModule,
    IdentityProviderModule,
    InternationalizationModule,
    ServiceWorkerModule.register('ngsw-worker.js', { enabled: SERVICE_WORKER }),
    StateManagementModule,
  ],
  providers: [
    { provide: HTTP_INTERCEPTORS, useClass: ICMErrorMapperInterceptor, multi: true },
    {
      provide: HTTP_INTERCEPTORS,
      useClass: IdentityProviderInterceptor,
      multi: true,
    },
    { provide: HTTP_INTERCEPTORS, useClass: MockInterceptor, multi: true },
    { provide: ErrorHandler, useClass: DefaultErrorHandler },
    {
      provide: APP_BASE_HREF,
      useFactory: (s: PlatformLocation, baseHref: string) => baseHref || s.getBaseHrefFromDOM(),
      deps: [PlatformLocation, [new Optional(), new SkipSelf(), APP_BASE_HREF]],
    },
  ],
  // exports needed to use the cookie banner in the AppComponent
  exports: [TranslateModule],
})
export class CoreModule {
  constructor(@Optional() @SkipSelf() parentModule: CoreModule) {
    if (parentModule) {
      throw new Error('CoreModule is already loaded. Import it in the AppModule only');
    }
  }
}
