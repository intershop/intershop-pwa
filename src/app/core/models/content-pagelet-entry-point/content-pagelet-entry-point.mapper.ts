import { Injectable } from '@angular/core';

import { ContentConfigurationParameterMapper } from 'ish-core/models/content-configuration-parameter/content-configuration-parameter.mapper';
import { ContentPageletEntryPointData } from 'ish-core/models/content-pagelet-entry-point/content-pagelet-entry-point.interface';
import { ContentPageletEntryPoint } from 'ish-core/models/content-pagelet-entry-point/content-pagelet-entry-point.model';
import { ContentPageletMapper } from 'ish-core/models/content-pagelet/content-pagelet.mapper';
import { ContentPagelet } from 'ish-core/models/content-pagelet/content-pagelet.model';

@Injectable({ providedIn: 'root' })
export class ContentPageletEntryPointMapper {
  constructor(
    private contentConfigurationParameterMapper: ContentConfigurationParameterMapper,
    private contentPageletMapper: ContentPageletMapper
  ) {}

  /**
   * Converts {@link ContentPageletEntryPointData} to the model entity {@link ContentPageletEntryPoint} and enclosed {@link ContentPagelet}s.
   */
  fromData(data: ContentPageletEntryPointData): [ContentPageletEntryPoint, ContentPagelet[]] {
    if (!data) {
      throw new Error('falsy input');
    }

    let pagelets: ContentPagelet[] = [];
    let pageletIDs: string[] = [];

    if (data.pagelets) {
      pageletIDs = data.pagelets.map(p => p.id);
      pagelets = data.pagelets
        .map(x => this.contentPageletMapper.fromData(x))
        .reduce((acc, val) => [...acc, ...val], []);
    }

    const configurationParameters = this.contentConfigurationParameterMapper.fromData(data.configurationParameters);

    const pageletEntryPoint: ContentPageletEntryPoint = {
      id: data.id,
      definitionQualifiedName: data.definitionQualifiedName,
      displayName: data.displayName,
      domain: data.domain,
      resourceSetId: data.resourceSetId,
      pageletIDs,
      configurationParameters,
    };

    return [pageletEntryPoint, pagelets];
  }
}
