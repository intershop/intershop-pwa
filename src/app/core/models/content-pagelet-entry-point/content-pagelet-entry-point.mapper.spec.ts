import { TestBed } from '@angular/core/testing';
import { provideMockStore } from '@ngrx/store/testing';

import { ContentPageletEntryPointData } from 'ish-core/models/content-pagelet-entry-point/content-pagelet-entry-point.interface';
import { ContentPageletEntryPointMapper } from 'ish-core/models/content-pagelet-entry-point/content-pagelet-entry-point.mapper';

describe('Content Pagelet Entry Point Mapper', () => {
  let contentPageletEntryPointMapper: ContentPageletEntryPointMapper;

  beforeEach(() => {
    TestBed.configureTestingModule({
      providers: [provideMockStore()],
    });
    contentPageletEntryPointMapper = TestBed.inject(ContentPageletEntryPointMapper);
  });

  it('should throw on falsy input', () => {
    expect(() => contentPageletEntryPointMapper.fromData(undefined)).toThrowError('falsy input');
  });

  it('should convert a complex example to complex type', () => {
    const input: ContentPageletEntryPointData = {
      link: {
        title: 'Readable Title',
        uri: 'uri://test',
        type: 'ContentInclude',
      },
      id: 'include-id',
      domain: 'the-domain',
      resourceSetId: 'resource-set-id',
      definitionQualifiedName: 'fq',
      displayName: 'name-include',
      pagelets: [
        {
          id: 'pagelet1',
          definitionQualifiedName: 'fq',
          domain: 'domain',
          displayName: 'name1',
          configurationParameters: {
            key1: {
              definitionQualifiedName: 'fq',
              value: 'hallo 1',
            },
          },
          slots: {
            slot1: {
              displayName: 'slot1',
              definitionQualifiedName: 'fq',
              pagelets: [
                {
                  id: 'pagelet11',
                  definitionQualifiedName: 'fq',
                  domain: 'Domain',
                  displayName: 'name11',
                  configurationParameters: {
                    key11: {
                      definitionQualifiedName: 'fq',
                      value: 'test',
                    },
                  },
                },
              ],
            },
          },
        },
        {
          id: 'pagelet2',
          definitionQualifiedName: 'fq',
          domain: 'domain',
          displayName: 'name2',
        },
      ],
    };

    const [pageletEntryPoint, pagelets] = contentPageletEntryPointMapper.fromData(input);

    expect(pagelets).toHaveLength(3);
    expect(pagelets.map(p => p.id)).toIncludeAllMembers(['pagelet1', 'pagelet11', 'pagelet2']);
    expect(pageletEntryPoint).not.toBeEmpty();
    expect(pageletEntryPoint).toMatchInlineSnapshot(`
      Object {
        "configurationParameters": Object {},
        "definitionQualifiedName": "fq",
        "displayName": "name-include",
        "domain": "the-domain",
        "id": "include-id",
        "pageletIDs": Array [
          "pagelet1",
          "pagelet2",
        ],
        "resourceSetId": "resource-set-id",
      }
    `);
  });
});
