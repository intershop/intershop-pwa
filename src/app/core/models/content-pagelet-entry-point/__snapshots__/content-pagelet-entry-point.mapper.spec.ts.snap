// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Content Pagelet Entry Point Mapper should convert a complex example to complex type 1`] = `
Object {
  "pageletEntryPoint": Object {
    "clientId": "include-id",
    "configurationParameters": Object {},
    "definitionQualifiedName": "fq",
    "displayName": "name-include",
    "domain": "the-domain",
    "id": "include-id",
    "pageletIDs": Array [
      "pagelet1",
      "pagelet2",
    ],
    "resourceSetId": "resource-set-id",
  },
  "pagelets": Array [
    Object {
      "configurationParameters": Object {
        "key1": "hallo 1",
      },
      "definitionQualifiedName": "fq",
      "displayName": "name1",
      "domain": "domain",
      "id": "pagelet1",
      "slots": Array [
        Object {
          "configurationParameters": Object {},
          "definitionQualifiedName": "fq",
          "displayName": "slot1",
          "pageletIDs": Array [
            "pagelet11",
          ],
        },
      ],
    },
    Object {
      "configurationParameters": Object {
        "key11": "test",
      },
      "definitionQualifiedName": "fq",
      "displayName": "name11",
      "domain": "Domain",
      "id": "pagelet11",
      "slots": Array [],
    },
    Object {
      "configurationParameters": Object {},
      "definitionQualifiedName": "fq",
      "displayName": "name2",
      "domain": "domain",
      "id": "pagelet2",
      "slots": Array [],
    },
  ],
}
`;
